feature: implemented a user flags attribute.

thibaultcha opened the pull request:
Like `lua_shared_dict`'s `flags` attribute, but for the Lua-land. This would be extremely useful to applications who wish to store some metadata alongside scalar cached values. I have a need for this in a library already.

thibaultcha commented on lib/resty/lrucache.lua:
```diff
@@ -237,6 +244,12 @@ function _M.set(self, key, value, ttl)
     end
+
+    if flags then
+        node.user_flags = ffi_new(flags_type, flags)
```
When implementing this I was wondering if instead it shoud be preferred to pre-allocate it at queue init time, and use a flag to signify whether the `user_flags` field is relevant or not. Please advise if this is fine or not.

dndx replied:
I think we should just embed `flag` in the struct itself. Allocating separately doesn't really save us any space anyway and is definitely slower.

thibaultcha replied:
@dndx Agreed, that was my feeling as well; done!

thibaultcha commented:
@dndx @agentzh Hey there; do you have some time for a review here? This lib has some pending PRs that feel ready and could be quite useful :)

agentzh commented on t/mixed.t:
```diff
@@ -39,8 +39,8 @@ __DATA__
-            ngx.say("dog: ", c:get("dog"))
-            ngx.say("cat: ", c:get("cat"))
+            ngx.say("dog: ", (c:get("dog")))
+            ngx.say("cat: ", (c:get("cat")))
```
Seems like this change may break existing user Lua code like this?

thibaultcha replied:
That seems unlikely? I would theorize that most users of this library would already use it in its intended form (without even using `stale_value`):
```lua
local value = c:get("key")
```

It could be more surprising when used in tail-calls:
```lua
return c:get("key")
```

but then we simply return more values that will likely simply be discarded. _Maybe_ somebody uses:
```lua
return c:get("key"), ret3, ret4
```

But it would be a shame to miss on this feature and a chance to align with the shm API for such exotic usage, no? I can add an argument `c:get(key, flags?)` or implement this in `value, stale_value, flags = c:get_flags(key)` if you prefer to be conservative.

dndx replied:
Revisiting this PR after a long time...

We do this already in shared dict. Realistically no user is really gonna use the library by passing the returns directly to `ngx.say`, I think it's just a matter of tests not being written properly (which is also specific to this particular repository).

I think between signed integers and bit fields, sticking with what shdict does, that is, setting `flags` to `0` by default makes most sense and aligns with the rest of OpenResty convention the most.

agentzh commented on lib/resty/lrucache.lua:
```diff
@@ -169,11 +171,16 @@ function _M.get(self, key)
+        return nil, val, flags
     end
-    return val
+    return val, nil, flags
```
Better not return nil flag values to minimize API compatibility changes.

thibaultcha replied:
Yeah, that sounds sane, and I was considering it when implementing this as well.

thibaultcha replied:
@agentzh Updated.

dndx commented on lib/resty/lrucache/pureffi.lua:
```diff
@@ -165,6 +166,8 @@ ffi.cdef[[
 
         double             expire;  /* in seconds */
 
+        unsigned int      *user_flags;
```
I think all field should align so maybe we can move `expire` and `*user_flags` backwards together.

thibaultcha commented:
@dndx @agentzh If you have time for another look,  everything so far has been addressed.

thibaultcha commented:
Ping.

agentzh commented on README.markdown:
```diff
@@ -155,24 +155,33 @@ to the range of `[0.1, 1]` (i.e. if load factor is greater than 1, it will be sa
 
+The optional `flags` argument specifies a user flags value associated with the
+item to be stored. It can be retrieved later with the item. The user
+flags is stored as an unsigned 32-bit integer internally, and thus must be
```
`flags` is a plural word and it looks weird to be used with `is`.

agentzh commented on lib/resty/lrucache.lua:
```diff
@@ -39,13 +40,16 @@ ffi.cdef[[
+        unsigned int      *user_flags;
         lrucache_queue_t  *prev;
         lrucache_queue_t  *next;
+        unsigned int       has_flags:1;
```
Should put `:1` into a comment, just like the nginx core's code base. Basically if there is only a single bit field in the end, then we should not use bit field for it since it only slows things down.

Actually I think it's more CPU efficient to directly store the flags here.

agentzh commented on lib/resty/lrucache.lua:
```diff
@@ -75,6 +79,7 @@ local function queue_init(size)
         local prev = q[0]
         for i = 1, size do
           local e = q + i
+          e.user_flags = ffi_new(flags_type)
```
This creates too many small cdata objects and add too much GC overhead. We should either preallocate it or just add the flags to the node data (which is preferred).

thibaultcha replied:
It is already pre-allocated per lrucache instance. Unless you mean something else? I am not sure where else to store the data, this is, afaict, the node data already. Do you maybe mean another dictionary attribute `self.hash_flags`, similar to `self.hasht`?

Like so maybe?
```lua
self.hasht = value
self.hash_flags[key] = flags
```

thibaultcha replied:
@agentzh I updated the patch. Is that what you meant?

dndx commented on lib/resty/lrucache/pureffi.lua:
```diff
@@ -528,6 +539,13 @@ function _M.set(self, key, value, ttl)
+        node.user_flags = flags
+
+    else
+        node.user_flags = -1
```
Second thought on this, we should probably do what shdict does and sets the default flag value to `0` instead of `-1`. Otherwise we can not really support "32-bit flags" as `-1` also coincident with all bits being turned on.

dndx replied:
This also means we may want to store `flags` as `uint32_t` instead of a signed integer, same as shdict.

thibaultcha commented:
PR rebased and updated. I reverted it back to the "always return flags" behaviour, and introduced the test suite changes again (ensuring that `ngx.say` only sees one return value).

dndx commented on lib/resty/lrucache.lua:
```diff
@@ -42,6 +43,7 @@ ffi.cdef[[
         double             expire;  /* in seconds */
         lrucache_queue_t  *prev;
         lrucache_queue_t  *next;
+        unsigned int       user_flags;
```
Let's use `uint32_t` here instead. I know almost all modern architecture has 4 byte `int`s, but it's better to be explicit.

dndx commented on README.markdown:
```diff
@@ -155,24 +155,33 @@ to the range of `[0.1, 1]` (i.e. if load factor is greater than 1, it will be sa
 
+Starting from `v0.10`, the user flags value associated with the stored item is
+also returned as the third return value. If no user flags were given to an
+item, this return value will be `nil`.
```
This should be the same as shdict, that is, the default flags is `0` and will be returned if not specified.

thibaultcha replied:
Oh, I forgot to update the documentation yesterday, thanks.

dndx commented on README.markdown:
```diff
@@ -155,24 +155,33 @@ to the range of `[0.1, 1]` (i.e. if load factor is greater than 1, it will be sa
 set
 ---
-`syntax: cache:set(key, value, ttl)`
+`syntax: cache:set(key, value, ttl, flags)`
```
Should be `flags?` to signal it is optional.

dndx commented on lib/resty/lrucache/pureffi.lua:
```diff
@@ -156,14 +158,16 @@ ffi.cdef[[
-        int                conflict;
+        int                        conflict;
+
+        unsigned int               user_flags;
```
Same here.

dndx commented on lib/resty/lrucache/pureffi.lua:
```diff
@@ -70,6 +70,8 @@ local uintptr_t = ffi.typeof("uintptr_t")
 local c_str_t = ffi.typeof("const char*")
 local int_t = ffi.typeof("int")
 local int_array_t = ffi.typeof("int[?]")
+local flags_type = ffi.typeof("unsigned int[1]")
```
Is this still being used? I can't seems to find `flags_type` anywhere else in this file.

thibaultcha replied:
Also removed the below `type` which was already localized below.

dndx commented on README.markdown:
```diff
@@ -155,24 +155,33 @@ to the range of `[0.1, 1]` (i.e. if load factor is greater than 1, it will be sa
+The optional `flags` argument specifies a user flags value associated with the
+item to be stored. It can be retrieved later with the item. The user flags are
+stored as an unsigned 32-bit integer internally, and thus must be specified as
+a Lua number. This argument was added in the `v0.10` release.
```
Better add "if not specified, the flag will be set to `0`" in here as well.

thibaultcha replied:
Didn't judge it necessary (since it has no impact on the user code until `get()`, but added it now, no strong feelings.

thibaultcha commented:
We should not merge this until after the 1.15 formal release anyway, as the merge window should still be in effect in case of bugs in this library, so no rush. Thanks for the review @dndx!
